File: examples-ipc/go/service.go
package main

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net"
	"os"
	"time"
)

// MessageType 消息类型
type MessageType byte

const (
	REQUEST   MessageType = 1
	RESPONSE  MessageType = 2
	REGISTER  MessageType = 3
	HEARTBEAT MessageType = 4
)

// Message IPC消息结构
type Message struct {
	Type     MessageType
	ID       string
	Service  string
	Method   string
	Metadata map[string]string
	Data     []byte
}

// Handler 处理函数类型
type Handler func(params map[string]interface{}) (interface{}, error)

// NeoIPCClient IPC客户端
type NeoIPCClient struct {
	conn        net.Conn
	serviceName string
	handlers    map[string]Handler
}

// NewNeoIPCClient 创建IPC客户端
func NewNeoIPCClient(addr string) (*NeoIPCClient, error) {
	conn, err := net.Dial("tcp", addr)
	if err != nil {
		return nil, err
	}

	log.Printf("Connected to Neo IPC server at %s", addr)

	return &NeoIPCClient{
		conn:     conn,
		handlers: make(map[string]Handler),
	}, nil
}

// RegisterService 注册服务
func (c *NeoIPCClient) RegisterService(name string, metadata map[string]string) error {
	c.serviceName = name

	registerData := map[string]interface{}{
		"name":     name,
		"metadata": metadata,
	}

	data, _ := json.Marshal(registerData)

	msg := &Message{
		Type:     REGISTER,
		ID:       "",
		Service:  name,
		Method:   "",
		Metadata: map[string]string{},
		Data:     data,
	}

	if err := c.sendMessage(msg); err != nil {
		return err
	}

	log.Printf("Service '%s' registered", name)
	return nil
}

// AddHandler 添加处理器
func (c *NeoIPCClient) AddHandler(method string, handler Handler) {
	c.handlers[method] = handler
	log.Printf("Handler registered for method: %s", method)
}

// sendMessage 发送消息
func (c *NeoIPCClient) sendMessage(msg *Message) error {
	buf := &bytes.Buffer{}

	// 消息类型
	buf.WriteByte(byte(msg.Type))

	// ID
	binary.Write(buf, binary.LittleEndian, uint32(len(msg.ID)))
	buf.WriteString(msg.ID)

	// Service
	binary.Write(buf, binary.LittleEndian, uint32(len(msg.Service)))
	buf.WriteString(msg.Service)

	// Method
	binary.Write(buf, binary.LittleEndian, uint32(len(msg.Method)))
	buf.WriteString(msg.Method)

	// Metadata
	metadataJSON, _ := json.Marshal(msg.Metadata)
	binary.Write(buf, binary.LittleEndian, uint32(len(metadataJSON)))
	buf.Write(metadataJSON)

	// Data
	binary.Write(buf, binary.LittleEndian, uint32(len(msg.Data)))
	buf.Write(msg.Data)

	// 发送消息长度和内容
	content := buf.Bytes()
	binary.Write(c.conn, binary.LittleEndian, uint32(len(content)))
	_, err := c.conn.Write(content)

	return err
}

// readMessage 读取消息
func (c *NeoIPCClient) readMessage() (*Message, error) {
	// 读取消息长度
	var msgLen uint32
	if err := binary.Read(c.conn, binary.LittleEndian, &msgLen); err != nil {
		return nil, err
	}

	// 读取消息内容
	msgData := make([]byte, msgLen)
	if _, err := io.ReadFull(c.conn, msgData); err != nil {
		return nil, err
	}

	reader := bytes.NewReader(msgData)
	msg := &Message{}

	// 解析消息类型
	var msgType byte
	binary.Read(reader, binary.LittleEndian, &msgType)
	msg.Type = MessageType(msgType)

	// 解析ID
	var idLen uint32
	binary.Read(reader, binary.LittleEndian, &idLen)
	idBytes := make([]byte, idLen)
	io.ReadFull(reader, idBytes)
	msg.ID = string(idBytes)

	// 解析Service
	var serviceLen uint32
	binary.Read(reader, binary.LittleEndian, &serviceLen)
	serviceBytes := make([]byte, serviceLen)
	io.ReadFull(reader, serviceBytes)
	msg.Service = string(serviceBytes)

	// 解析Method
	var methodLen uint32
	binary.Read(reader, binary.LittleEndian, &methodLen)
	methodBytes := make([]byte, methodLen)
	io.ReadFull(reader, methodBytes)
	msg.Method = string(methodBytes)

	// 解析Metadata
	var metadataLen uint32
	binary.Read(reader, binary.LittleEndian, &metadataLen)
	metadataBytes := make([]byte, metadataLen)
	io.ReadFull(reader, metadataBytes)
	if len(metadataBytes) > 0 {
		json.Unmarshal(metadataBytes, &msg.Metadata)
	} else {
		msg.Metadata = make(map[string]string)
	}

	// 解析Data
	var dataLen uint32
	binary.Read(reader, binary.LittleEndian, &dataLen)
	msg.Data = make([]byte, dataLen)
	io.ReadFull(reader, msg.Data)

	return msg, nil
}

// handleRequest 处理请求
func (c *NeoIPCClient) handleRequest(msg *Message) {
	handler, ok := c.handlers[msg.Method]
	if !ok {
		errorResp := &Message{
			Type:    RESPONSE,
			ID:      msg.ID,
			Service: msg.Service,
			Method:  msg.Method,
			Metadata: map[string]string{
				"error": "true",
			},
			Data: []byte(fmt.Sprintf(`{"error":"Method '%s' not found"}`, msg.Method)),
		}
		c.sendMessage(errorResp)
		return
	}

	// 解析请求参数
	var params map[string]interface{}
	if len(msg.Data) > 0 {
		json.Unmarshal(msg.Data, &params)
	}

	// 调用处理器
	result, err := handler(params)
	if err != nil {
		errorResp := &Message{
			Type:    RESPONSE,
			ID:      msg.ID,
			Service: msg.Service,
			Method:  msg.Method,
			Metadata: map[string]string{
				"error": "true",
			},
			Data: []byte(fmt.Sprintf(`{"error":"%s"}`, err.Error())),
		}
		c.sendMessage(errorResp)
		return
	}

	// 发送响应
	responseData, _ := json.Marshal(result)
	response := &Message{
		Type:     RESPONSE,
		ID:       msg.ID,
		Service:  msg.Service,
		Method:   msg.Method,
		Metadata: map[string]string{},
		Data:     responseData,
	}
	c.sendMessage(response)
}

// heartbeatLoop 心跳循环
func (c *NeoIPCClient) heartbeatLoop() {
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()

	for range ticker.C {
		msg := &Message{
			Type:     HEARTBEAT,
			ID:       "",
			Service:  c.serviceName,
			Method:   "",
			Metadata: map[string]string{},
			Data:     []byte{},
		}
		if err := c.sendMessage(msg); err != nil {
			log.Printf("Heartbeat error: %v", err)
			return
		}
		log.Println("Heartbeat sent")
	}
}

// Run 运行服务
func (c *NeoIPCClient) Run() error {
	// 启动心跳
	go c.heartbeatLoop()

	// 消息处理循环
	for {
		msg, err := c.readMessage()
		if err != nil {
			return err
		}

		if msg.Type == REQUEST {
			go c.handleRequest(msg)
		}
	}
}

func main() {
	// 从环境变量读取配置
	host := os.Getenv("NEO_IPC_HOST")
	if host == "" {
		host = "localhost"
	}
	port := os.Getenv("NEO_IPC_PORT")
	if port == "" {
		port = "29999"  // 使用正确的默认端口
	}

	// 创建客户端
	client, err := NewNeoIPCClient(fmt.Sprintf("%s:%s", host, port))
	if err != nil {
		log.Fatal(err)
	}

	// 注册处理器
	client.AddHandler("hello", func(params map[string]interface{}) (interface{}, error) {
		name, _ := params["name"].(string)
		if name == "" {
			name = "World"
		}
		return map[string]interface{}{
			"message":   fmt.Sprintf("Hello, %s!", name),
			"timestamp": time.Now().Format(time.RFC3339),
			"service":   "Go Demo Service",
		}, nil
	})

	client.AddHandler("calculate", func(params map[string]interface{}) (interface{}, error) {
		a, _ := params["a"].(float64)
		b, _ := params["b"].(float64)
		operation, _ := params["operation"].(string)

		var result interface{}
		switch operation {
		case "add":
			result = a + b
		case "subtract":
			result = a - b
		case "multiply":
			result = a * b
		case "divide":
			if b != 0 {
				result = a / b
			} else {
				result = "Cannot divide by zero"
			}
		default:
			result = "Unknown operation"
		}

		return map[string]interface{}{
			"result":    result,
			"operation": operation,
			"a":         a,
			"b":         b,
		}, nil
	})

	client.AddHandler("echo", func(params map[string]interface{}) (interface{}, error) {
		message, _ := params["message"].(string)
		return map[string]interface{}{
			"echo":     message,
			"length":   len(message),
			"reversed": reverseString(message),
		}, nil
	})

	client.AddHandler("getTime", func(params map[string]interface{}) (interface{}, error) {
		format, _ := params["format"].(string)
		now := time.Now()

		var timeStr string
		switch format {
		case "unix":
			timeStr = fmt.Sprintf("%d", now.Unix())
		case "readable":
			timeStr = now.Format("2006-01-02 15:04:05")
		default:
			timeStr = now.Format(time.RFC3339)
		}

		return map[string]interface{}{
			"time":     timeStr,
			"timezone": now.Location().String(),
			"format":   format,
		}, nil
	})

	client.AddHandler("getInfo", func(params map[string]interface{}) (interface{}, error) {
		handlers := make([]string, 0, len(client.handlers))
		for method := range client.handlers {
			handlers = append(handlers, method)
		}

		return map[string]interface{}{
			"service":  "demo-service",
			"language": "Go",
			"version":  "1.0.0",
			"handlers": handlers,
			"uptime":   "N/A",
			"system": map[string]interface{}{
				"platform":   "go",
				"go_version": "1.x",
			},
		}, nil
	})

	// 注册服务
	err = client.RegisterService("demo-service", map[string]string{
		"language":    "go",
		"version":     "1.0.0",
		"description": "Go demo service for Neo Framework",
	})
	if err != nil {
		log.Fatal(err)
	}

	log.Println("Go demo service is running...")
	log.Printf("Listening on %s:%s", host, port)
	log.Println("Available methods: hello, calculate, echo, getTime, getInfo")

	// 运行服务
	if err := client.Run(); err != nil {
		log.Fatal(err)
	}
}

// reverseString 反转字符串
func reverseString(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

File: examples-ipc/java/Service.java
import java.io.*;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.concurrent.*;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

public class Service {
    private static final byte REQUEST = 1;
    private static final byte RESPONSE = 2;
    private static final byte REGISTER = 3;
    private static final byte HEARTBEAT = 4;
    
    private final Socket socket;
    private final DataInputStream input;
    private final DataOutputStream output;
    private final String serviceName = "demo-service-java";
    private final Map<String, Handler> handlers = new HashMap<>();
    private final Gson gson = new Gson();
    private final ExecutorService executor = Executors.newCachedThreadPool();
    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
    
    public interface Handler {
        Object handle(Map<String, Object> params) throws Exception;
    }
    
    public Service(String host, int port) throws IOException {
        this.socket = new Socket(host, port);
        this.input = new DataInputStream(socket.getInputStream());
        this.output = new DataOutputStream(socket.getOutputStream());
        System.out.println("Connected to Neo IPC server at " + host + ":" + port);
    }
    
    public void registerService(Map<String, String> metadata) throws IOException {
        Map<String, Object> registerData = new HashMap<>();
        registerData.put("name", serviceName);
        registerData.put("metadata", metadata);
        
        byte[] data = gson.toJson(registerData).getBytes();
        sendMessage(REGISTER, "", serviceName, "", new HashMap<>(), data);
        System.out.println("Service '" + serviceName + "' registered");
    }
    
    public void addHandler(String method, Handler handler) {
        handlers.put(method, handler);
        System.out.println("Handler registered for method: " + method);
    }
    
    private void sendMessage(byte type, String id, String service, String method,
                           Map<String, String> metadata, byte[] data) throws IOException {
        ByteArrayOutputStream content = new ByteArrayOutputStream();
        
        // 消息类型
        content.write(type);
        
        // ID
        byte[] idBytes = id.getBytes();
        writeInt(content, idBytes.length);
        content.write(idBytes);
        
        // Service
        byte[] serviceBytes = service.getBytes();
        writeInt(content, serviceBytes.length);
        content.write(serviceBytes);
        
        // Method
        byte[] methodBytes = method.getBytes();
        writeInt(content, methodBytes.length);
        content.write(methodBytes);
        
        // Metadata
        byte[] metadataBytes = gson.toJson(metadata).getBytes();
        writeInt(content, metadataBytes.length);
        content.write(metadataBytes);
        
        // Data
        writeInt(content, data.length);
        content.write(data);
        
        // 发送消息
        byte[] contentBytes = content.toByteArray();
        synchronized (output) {
            writeInt(output, contentBytes.length);
            output.write(contentBytes);
            output.flush();
        }
    }
    
    private void writeInt(OutputStream out, int value) throws IOException {
        ByteBuffer buffer = ByteBuffer.allocate(4);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        buffer.putInt(value);
        out.write(buffer.array());
    }
    
    private int readInt(InputStream in) throws IOException {
        byte[] bytes = new byte[4];
        in.read(bytes);
        ByteBuffer buffer = ByteBuffer.wrap(bytes);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        return buffer.getInt();
    }
    
    private Message readMessage() throws IOException {
        // 读取消息长度
        int msgLen = readInt(input);
        
        // 读取消息内容
        byte[] msgData = new byte[msgLen];
        input.readFully(msgData);
        
        ByteArrayInputStream bais = new ByteArrayInputStream(msgData);
        
        // 解析消息
        Message msg = new Message();
        msg.type = (byte) bais.read();
        
        // ID
        int idLen = readInt(bais);
        byte[] idBytes = new byte[idLen];
        bais.read(idBytes);
        msg.id = new String(idBytes);
        
        // Service
        int serviceLen = readInt(bais);
        byte[] serviceBytes = new byte[serviceLen];
        bais.read(serviceBytes);
        msg.service = new String(serviceBytes);
        
        // Method
        int methodLen = readInt(bais);
        byte[] methodBytes = new byte[methodLen];
        bais.read(methodBytes);
        msg.method = new String(methodBytes);
        
        // Metadata
        int metadataLen = readInt(bais);
        byte[] metadataBytes = new byte[metadataLen];
        bais.read(metadataBytes);
        if (metadataLen > 0) {
            msg.metadata = gson.fromJson(new String(metadataBytes), 
                new TypeToken<Map<String, String>>(){}.getType());
        } else {
            msg.metadata = new HashMap<>();
        }
        
        // Data
        int dataLen = readInt(bais);
        msg.data = new byte[dataLen];
        bais.read(msg.data);
        
        return msg;
    }
    
    private void handleRequest(Message msg) {
        Handler handler = handlers.get(msg.method);
        if (handler == null) {
            try {
                Map<String, String> errorMeta = new HashMap<>();
                errorMeta.put("error", "true");
                String errorData = String.format("{\"error\":\"Method '%s' not found\"}", msg.method);
                sendMessage(RESPONSE, msg.id, msg.service, msg.method, errorMeta, errorData.getBytes());
            } catch (IOException e) {
                e.printStackTrace();
            }
            return;
        }
        
        try {
            // 解析请求参数
            Map<String, Object> params = new HashMap<>();
            if (msg.data.length > 0) {
                params = gson.fromJson(new String(msg.data), 
                    new TypeToken<Map<String, Object>>(){}.getType());
            }
            
            // 调用处理器
            Object result = handler.handle(params);
            
            // 发送响应
            byte[] responseData = gson.toJson(result).getBytes();
            sendMessage(RESPONSE, msg.id, msg.service, msg.method, new HashMap<>(), responseData);
            
        } catch (Exception e) {
            try {
                Map<String, String> errorMeta = new HashMap<>();
                errorMeta.put("error", "true");
                String errorData = String.format("{\"error\":\"%s\"}", e.getMessage());
                sendMessage(RESPONSE, msg.id, msg.service, msg.method, errorMeta, errorData.getBytes());
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
    }
    
    private void startHeartbeat() {
        scheduler.scheduleAtFixedRate(() -> {
            try {
                sendMessage(HEARTBEAT, "", serviceName, "", new HashMap<>(), new byte[0]);
                System.out.println("Heartbeat sent");
            } catch (IOException e) {
                e.printStackTrace();
            }
        }, 30, 30, TimeUnit.SECONDS);
    }
    
    public void run() throws IOException {
        startHeartbeat();
        
        while (!socket.isClosed()) {
            Message msg = readMessage();
            if (msg.type == REQUEST) {
                executor.execute(() -> handleRequest(msg));
            }
        }
    }
    
    private static class Message {
        byte type;
        String id;
        String service;
        String method;
        Map<String, String> metadata;
        byte[] data;
    }
    
    public static void main(String[] args) throws Exception {
        // 从环境变量读取配置
        String host = System.getenv("NEO_IPC_HOST");
        if (host == null) host = "localhost";
        
        String portStr = System.getenv("NEO_IPC_PORT");
        int port = portStr != null ? Integer.parseInt(portStr) : 9999;  // 使用正确的默认端口
        
        // 创建服务
        Service service = new Service(host, port);
        
        // 注册处理器
        service.addHandler("hello", params -> {
            String name = (String) params.getOrDefault("name", "World");
            Map<String, Object> response = new HashMap<>();
            response.put("message", "Hello, " + name + "!");
            response.put("timestamp", new Date().toString());
            response.put("service", "Java Demo Service");
            return response;
        });
        
        service.addHandler("calculate", params -> {
            double a = ((Number) params.getOrDefault("a", 0)).doubleValue();
            double b = ((Number) params.getOrDefault("b", 0)).doubleValue();
            String operation = (String) params.getOrDefault("operation", "add");
            
            Object result;
            switch (operation) {
                case "add":
                    result = a + b;
                    break;
                case "subtract":
                    result = a - b;
                    break;
                case "multiply":
                    result = a * b;
                    break;
                case "divide":
                    result = b != 0 ? a / b : "Cannot divide by zero";
                    break;
                default:
                    result = "Unknown operation";
            }
            
            Map<String, Object> response = new HashMap<>();
            response.put("result", result);
            response.put("operation", operation);
            response.put("a", a);
            response.put("b", b);
            return response;
        });
        
        service.addHandler("echo", params -> {
            String message = (String) params.getOrDefault("message", "");
            Map<String, Object> response = new HashMap<>();
            response.put("echo", message);
            response.put("length", message.length());
            response.put("reversed", new StringBuilder(message).reverse().toString());
            return response;
        });
        
        service.addHandler("getTime", params -> {
            String format = (String) params.getOrDefault("format", "iso");
            Date now = new Date();
            
            String timeStr;
            switch (format) {
                case "unix":
                    timeStr = String.valueOf(now.getTime() / 1000);
                    break;
                case "readable":
                    timeStr = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(now);
                    break;
                default:
                    timeStr = now.toString();
            }
            
            Map<String, Object> response = new HashMap<>();
            response.put("time", timeStr);
            response.put("timezone", TimeZone.getDefault().getID());
            response.put("format", format);
            return response;
        });
        
        service.addHandler("getInfo", params -> {
            Map<String, Object> response = new HashMap<>();
            response.put("service", "demo-service-java");
            response.put("language", "Java");
            response.put("version", "1.0.0");
            response.put("handlers", new ArrayList<>(service.handlers.keySet()));
            response.put("uptime", "N/A");
            
            Map<String, String> system = new HashMap<>();
            system.put("platform", "java");
            system.put("java_version", System.getProperty("java.version"));
            response.put("system", system);
            
            return response;
        });
        
        // 注册服务
        Map<String, String> metadata = new HashMap<>();
        metadata.put("language", "java");
        metadata.put("version", "1.0.0");
        metadata.put("description", "Java demo service for Neo Framework");
        service.registerService(metadata);
        
        System.out.println("Java demo service is running...");
        System.out.println("Listening on " + host + ":" + port);
        System.out.println("Available methods: hello, calculate, echo, getTime, getInfo");
        
        // 运行服务
        service.run();
    }
}